### PR TITLE
wip: design: Header routing design doc

### PR DESCRIPTION
**Pull request by stevesloka**
Updates #947 by adding a Header based routing design document. 

Signed-off-by: Steve Sloka <slokas@vmware.com>

### CODE REVIEW

**Comment by stevesloka** [comment]
I have a question on if Contour should be able to route on **only** a Header or does it require a path to go with it. This design requires a path which I've outlined in the design.

File: design/header-routing.md [hunk 1]
```diff
@@ -0,0 +1,71 @@
+# Hewader Based Routing
```
Comment (pbarker) [inline design/header-routing.md]: sp

File: design/header-routing.md [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+The IngressRoute spec will be updated to allow for a `header` field to be added which will allow for a set of key/value pairs to be applied to the route.
+In the following example, you can define for path `/foo` and send traffic to different services for the same path but which contain different headers.
+
+```
+spec:
+  routes:
+  - match: /foo
+      header:
+        name: "x-header"
+        value: "A"
```
Comment (davecheney) [inline design/header-routing.md]: `value: "A"` here looks like an exact match. What about if I wanted to match on a prefix, or substring -- ie, the header value must contain a certain string.

File: design/header-routing.md [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    services:
+    - name: backend-A
+        port: 9999
+  - match: /foo
```
Comment (davecheney) [inline design/header-routing.md]: What about the use case where I only want to match on header only? Does it make sense to always have to combine this with a path?

File: design/header-routing.md (hunk 2 above)
Comment (stevesloka) [reply design/header-routing.md]: Are you suggesting something like "A*" or "*A*" or even "AAAA*AAAA"?

**Comment by stevesloka** [comment]
Closing this because the IngressRoutev1 design doc implements.

**Pull request closed without merging**